File: web-analyst-web/frontend/src/core/domain/chunking/sentenceChunker.ts
export interface ChunkMetadata {
  sourceId: string;
  chunkIndex: number;
  url?: string;
  title?: string;
}

export interface TextChunk {
  text: string;
  metadata: ChunkMetadata;
}

export interface ChunkingOptions {
  maxChars?: number;
  overlapSentences?: number;
}

function splitIntoSentences(text: string): string[] {
  // Try Intl.Segmenter if available (modern browsers)
  if (typeof Intl !== 'undefined' && 'Segmenter' in Intl) {
    try {
      const segmenter = new (Intl as any).Segmenter('en', { granularity: 'sentence' });
      const segments = segmenter.segment(text);
      return Array.from(segments, (s: any) => s.segment.trim()).filter((s: string) => s.length > 0);
    } catch (e) {
      // Fall through to regex
    }
  }

  // Regex fallback: split on sentence boundaries
  // Handles: . ! ? followed by space/newline, but not abbreviations like "Dr." or "U.S."
  const sentences = text
    .replace(/([.!?])\s+(?=[A-Z])/g, '$1|')
    .replace(/([.!?])(\n+)/g, '$1|')
    .split('|')
    .map(s => s.trim())
    .filter(s => s.length > 0);

  return sentences;
}

export function chunkText(
  text: string,
  sourceId: string,
  sourceUrl?: string,
  sourceTitle?: string,
  options: ChunkingOptions = {}
): TextChunk[] {
  const maxChars = options.maxChars || 1500;
  const overlapSentences = options.overlapSentences || 1;

  if (!text || text.trim().length === 0) {
    return [];
  }

  const sentences = splitIntoSentences(text);
  
  if (sentences.length === 0) {
    return [];
  }

  const chunks: TextChunk[] = [];
  let currentChunk: string[] = [];
  let currentLength = 0;
  let chunkIndex = 0;

  for (let i = 0; i < sentences.length; i++) {
    const sentence = sentences[i];
    const sentenceLength = sentence.length;

    // If single sentence exceeds maxChars, split it as its own chunk
    if (sentenceLength > maxChars && currentChunk.length === 0) {
      chunks.push({
        text: sentence,
        metadata: {
          sourceId,
          chunkIndex: chunkIndex++,
          url: sourceUrl,
          title: sourceTitle,
        },
      });
      continue;
    }

    // If adding this sentence would exceed maxChars, finalize current chunk
    if (currentLength + sentenceLength > maxChars && currentChunk.length > 0) {
      chunks.push({
        text: currentChunk.join(' '),
        metadata: {
          sourceId,
          chunkIndex: chunkIndex++,
          url: sourceUrl,
          title: sourceTitle,
        },
      });

      // Start new chunk with overlap
      const overlapStart = Math.max(0, currentChunk.length - overlapSentences);
      currentChunk = currentChunk.slice(overlapStart);
      currentLength = currentChunk.reduce((sum, s) => sum + s.length, 0);
    }

    currentChunk.push(sentence);
    currentLength += sentenceLength;
  }

  // Add final chunk if not empty
  if (currentChunk.length > 0) {
    chunks.push({
      text: currentChunk.join(' '),
      metadata: {
        sourceId,
        chunkIndex: chunkIndex++,
        url: sourceUrl,
        title: sourceTitle,
      },
    });
  }

  return chunks;
}

File: web-analyst-web/backend/src/domain/chunking/sentence_chunker.py
"""
Sentence-aware text chunking with metadata preservation.
Addresses mentor feedback: "Consider using sentence-aware chunking instead of word-based"
"""
import re
from typing import List, Dict, Optional
from dataclasses import dataclass


@dataclass
class ChunkMetadata:
    """Metadata for each chunk"""
    source_id: str
    chunk_index: int
    url: Optional[str] = None
    title: Optional[str] = None


@dataclass
class TextChunk:
    """Text chunk with metadata"""
    text: str
    metadata: ChunkMetadata


def split_into_sentences(text: str) -> List[str]:
    """
    Split text into sentences using regex with improved handling.
    
    Handles:
    - Standard sentence endings (. ! ?)
    - Abbreviations (Dr., U.S., etc.)
    - Multiple punctuation (!!, ?!)
    - Newlines as sentence boundaries
    """
    if not text or not text.strip():
        return []
    
    # Normalize whitespace and newlines
    text = re.sub(r'\n+', ' ', text)
    text = re.sub(r'\s+', ' ', text).strip()
    
    # Split on sentence boundaries: . ! ? followed by space and capital letter
    # Also handles multiple punctuation like !! or ?!
    sentences = re.split(r'(?<=[.!?])\s+(?=[A-Z])', text)
    
    # Filter out empty sentences and strip whitespace
    sentences = [s.strip() for s in sentences if s.strip()]
    
    return sentences


def chunk_text(
    text: str,
    source_id: str,
    url: Optional[str] = None,
    title: Optional[str] = None,
    max_chars: int = 1500,
    overlap_sentences: int = 1
) -> List[TextChunk]:
    """
    Chunk text using sentence-aware strategy.
    
    Args:
        text: Text to chunk
        source_id: Unique source identifier
        url: Source URL (optional)
        title: Source title (optional)
        max_chars: Maximum characters per chunk
        overlap_sentences: Number of sentences to overlap between chunks
        
    Returns:
        List of TextChunk objects with metadata
    """
    if not text or not text.strip():
        return []
    
    sentences = split_into_sentences(text)
    if not sentences:
        return []
    
    chunks: List[TextChunk] = []
    current_chunk: List[str] = []
    current_length = 0
    chunk_index = 0
    
    for sentence in sentences:
        sentence_length = len(sentence)
        
        # If single sentence exceeds max_chars, create its own chunk
        if sentence_length > max_chars and not current_chunk:
            chunks.append(TextChunk(
                text=sentence,
                metadata=ChunkMetadata(
                    source_id=source_id,
                    chunk_index=chunk_index,
                    url=url,
                    title=title
                )
            ))
            chunk_index += 1
            continue
        
        # If adding sentence would exceed max_chars, finalize current chunk
        if current_length + sentence_length > max_chars and current_chunk:
            chunks.append(TextChunk(
                text=' '.join(current_chunk),
                metadata=ChunkMetadata(
                    source_id=source_id,
                    chunk_index=chunk_index,
                    url=url,
                    title=title
                )
            ))
            chunk_index += 1
            
            # Start new chunk with overlap
            overlap_start = max(0, len(current_chunk) - overlap_sentences)
            current_chunk = current_chunk[overlap_start:]
            current_length = sum(len(s) for s in current_chunk)
        
        current_chunk.append(sentence)
        current_length += sentence_length
    
    # Add final chunk
    if current_chunk:
        chunks.append(TextChunk(
            text=' '.join(current_chunk),
            metadata=ChunkMetadata(
                source_id=source_id,
                chunk_index=chunk_index,
                url=url,
                title=title
            )
        ))
    
    return chunks
